### PR TITLE
Create `InterruptErrors` extension to raise an exception when an interrupted job is retried

### PR DESCRIPTION
**Pull request by bensheldon**
Connects to #821

### CODE REVIEW

**Comment by bensheldon** [comment]
@TAGraves for your review.

fyi, @jrochkind too because this is sort of related to #794

File: README.md [hunk 1]
```diff
@@ -829,6 +830,23 @@ end
...
+If you need more control over interrupt-driven retries, include the `InterruptErrors` extension in to your jobs. When an interrupted job is retried, the extension will raise an exception within the job, which allows you to use ActiveJob's `retry_on` and `discard_on` to control the behavior of the job.
+
+```ruby
+class MyJob < ApplicationJob
+  # The extension must be included before other extensions
+  include GoodJob::ActiveJobExtensions::InterruptErrors
+  # Discard the job if it is interrupted
+  discard_on InterruptedError
+  # Retry the job if it is interrupted
+  retry_on InterruptedError, wait: 0, attempts: Float::INFINITY
```
Comment (TAGraves) [inline README.md:846]: These should be `InterruptError`, not `InterruptedError`

**Review by TAGraves — APPROVED** [review]
🎉 This is going to be a huge help for us!

**Comment by bensheldon** [comment]
Released: https://github.com/bensheldon/good_job/releases/tag/v3.12.0

**Comment by jrochkind** [comment]
Thanks for the tag. I don't entirely understand what's going on, but seems good...

> Jobs will be automatically retried if the process is interrupted while performing a job, for example as the result of a `SIGKILL` or power failure.

Wait... is that a NEW feature? I thought that was always a feature of good_job from the start? No?

**Comment by bensheldon** [comment]
@jrochkind whoops, "will be" is just poor language on my part. It's "Jobs are automatically retried" which is how it's always been. 
 
#794 would log interrupted jobs that had been forcibly shutdown. This PR will, when those jobs are retried, provide some control over how they are executed (e.g. discard them rather than retry).

**Comment by jrochkind** [comment]
Thanks! Sounds good!

File: lib/good_job/active_job_extensions/interrupt_errors.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  module ActiveJobExtensions
+    module InterruptErrors
+      extend ActiveSupport::Concern
+
+      included do
+        around_perform do |_job, block|
+          raise InterruptError, "Interrupted after starting perform at '#{CurrentThread.execution_interrupted}'" if CurrentThread.execution_interrupted.present?
+
+          block.call
+        end
```
Comment (v2kovac) [inline lib/good_job/active_job_extensions/interrupt_errors.rb:12]: will this override/be overridden by any `around_perform` i have in my application_job.rb?

**Pull request merged**